python_package/madflow/makefile_template.py

```diff
@@ -0,0 +1,185 @@
+makefileName = "makefile"
+cppCompiler = "g++"
+cppVersion = "c++14"
+cudaPath = ""
+# example: cudaPath = "/usr/local/cuda"
+
+
+def write_compilers():
+    """Adds C++ and CUDA compilers"""
+    text = "CXX := " + cppCompiler + "\n"
+    text += "NVCC := $(shell which nvcc)\n"
+    text += "\n"
+    return text
+
+
+def write_shell_name():
+    """Adds a line for the kernel name"""
+    text = "UNAME_S := $(shell uname -s)\n"
+    text += "\n"
+    return text
+
+
+def write_multithreading():
+    """Find the number of processors and use as many threads as possible"""
+    text = "ifeq ($(UNAME_S), Darwin)\n"
+    text += "NPROCS = $(shell sysctl -n hw.ncpu)\n"
+    text += "else\n"
+    text += "NPROCS = $(shell grep -c 'processor' /proc/cpuinfo)\n"
+    text += "endif\n"
+    # if the number of processors isn't found, default to 1
+    text += "ifeq ($(NPROCS),)\n"
+    text += "NPROCS = 1\n"
+    text += "endif\n"
+    text += "MAKEFLAGS += -j$(NPROCS)\n"
+    text += "\n"
+    return text
+
+
+def write_tf_generic_flags():
+    """Adds TensorFlow flags"""
+    text = "TF_CFLAGS = $(shell python3 -c 'import tensorflow as tf; print(\" \".join(tf.sysconfig.get_compile_flags()))')\n"
+    text += "TF_LFLAGS = $(shell python3 -c 'import tensorflow as tf; print(\" \".join(tf.sysconfig.get_link_flags()))')\n"
+    text += "\n"
+    return text
+
+
+def write_tf_cuda_flags():
+    """Adds TansorFlow CUDA flags and the path to CUDA libraries.
+    If the environment variable ${CUDA_PATH} isn't defined, use a default path"""
+    text = "CUDA_LFLAGS = -x cu -Xcompiler -fPIC\n"
+    text += "CUDA_PATH := $(shell echo ${CUDA_PATH})\n"
+    text += "ifeq ($(CUDA_PATH),)\n"
+    """If the path for CUDA libraries isn't explicitly stated at the beginning of this file, find it from ${PATH}"""
+    if cudaPath == "":
+        text += 'CUDA_PATH = $(shell echo ${PATH} | sed -e "s&.*:\([^:]*cuda[^/]*\).*&\\1&g")\n'
+    else:
+        text += "CUDA_PATH = " + cudaPath + "\n"
+    text += "endif\n"
+    text += "\n"
+    return text
+
+
+def write_omp_flags():
+    """Adds flags for OpenMP parallelization"""
+    text = "ifeq ($(UNAME_S), Darwin)\n"
+    text += "OMP_CFLAGS = -Xpreprocessor -fopenmp -lomp\n"
+    text += "else\n"
+    text += "OMP_CFLAGS = -fopenmp\n"
+    text += "endif\n"
+    text += "\n"
+    return text
```

Comment:
This is a suggestion that if you want you can apply to this entire module or ignore completely. One way you can take advantage of the fact that you are using functions to generate the `makefile` is to add some flexibility to them, for instance:

```suggestion
def write_omp_flags(openmp_flag="-fopenmp"):
    """Adds flags for OpenMP parallelization"""
    text = f"""ifeq ($(UNAME_S), Darwin)
OMP_CFLAGS = -Xpreprocessor {openmp_flag} -lomp
else
OMP_CFLAGS = {openmp_flag}
endif
"""
    return text
```

For isntance, at the moment your code works with gcc but it would fail with the intel compiler (for which the flag is `-qopenmp`). Now, you are not using the intel compiler so you can't even test, but by writting the function in this more flexible manner it allows for easier future extensions.

I've used the `"""` `""""` syntax and the f-string instead of adding the `\n` because I feel it was cleaner here. It's not necessary always the case of course.

Reply:
Done. I also used more f-strings in `makefile_template.py`.